### PR TITLE
[easy][dynamo] Cleanup guard builder selection

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111723

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -255,8 +232,8 @@ def __str__(self):
         output += f"    Guarded Class Weakref: {self.guarded_class_weakref}\n"
         return output
 
-    def create(self, local_builder: GuardBuilderBase, global_builder: GuardBuilderBase):
-        return self.create_fn(self.source.select(local_builder, global_builder), self)
+    def create(self, builder: GuardBuilderBase):
+        return self.create_fn(builder, self)
```
Comment (voznesenskym) [inline torch/_guards.py:236]: could collapse but meh

**Review by jon-chuang — APPROVED** [review]
Yep, noticed that global guards may reference local vars, and vice versa (e.g. guard on global dict item indexed via local key). So really they should share the same context (access to both global and local scope).

**Review by jansel — APPROVED** [review]

**Comment by voznesenskym** [comment]
> Yep, noticed that global guards may reference local vars, and vice versa (e.g. guard on global dict item indexed via local key). So really they should share the same context (access to both global and local scope).

It's a tad more subtle. Local should have access to both, global only to global. We collapsed builders here (https://github.com/pytorch/pytorch/pull/111226) , but keeping scopes cleanly separate is still the right thing to do.

**Pull request closed without merging**